### PR TITLE
TSPS-276 Use rawls to submit/monitor workflow and add wds row

### PR DESCRIPTION
**Pull request by jsotobroad**
### Description 

Still in draft, currently requires manual changes to rawls openapi.yml to work

### Jira Ticket
https://broadworkbench.atlassian.net/browse/TSPS-276
https://broadworkbench.atlassian.net/browse/TSPS-277
https://broadworkbench.atlassian.net/browse/TSPS-294

### CODE REVIEW

**Comment by jsotobroad** [comment]
this will require https://github.com/broadinstitute/rawls/pull/2999 to be merged so we can pull the newest rawls api-docs.yaml with the fixes we need to get our objects to work (which is what im doing locally and why the build/tests in this pr are failing so far)

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/RunImputationGcpJobFlight.java [hunk 1]
```diff
@@ -50,5 +53,22 @@ public RunImputationGcpJobFlight(FlightMap inputParameters, Object beanBag) {
         new PrepareImputationInputsStep(
             flightBeanBag.getPipelinesService(), flightBeanBag.getImputationConfiguration()),
         dbRetryRule);
+
+    addStep(
+        new AddDataTableRowStep(flightBeanBag.getRawlsService(), flightBeanBag.getSamService()));
+
+    addStep(
+        new SubmitCromwellRunSetStep(
```
[the remark this answers is not in the document]
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/stairway/imputation/RunImputationGcpJobFlight.java:61]: oh you renamed the polling step to submissions - probably just want to rename this one too

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/PollCromwellSubmissionStatusStep.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    // if there are any non-successful workflows, fatally fail the step
+    List<Workflow> failedRunLogs =
+        submissionResponse.getWorkflows().stream()
+            .filter(workflow -> !workflow.getStatus().equals(WorkflowStatus.SUCCEEDED))
+            .toList();
+    if (failedRunLogs.isEmpty()) {
+      return StepResult.getStepResultSuccess();
+    } else {
+      return new StepResult(
+          StepStatus.STEP_RESULT_FAILURE_FATAL,
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/PollCromwellSubmissionStatusStep.java:95]: if we do a fatal failure here, do we need to increment the failed run counter here too? alternatively we can do a non-fatal failure so we run through the undo steps (i don't think we actually undo anything) so we get to the failed run counter increment that we will add to the undoStep method of whatever ends up being the first step.

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    // create submission request
+    SubmissionRequest submissionRequest =
+        new SubmissionRequest()
+            .entityName(flightContext.getFlightId())
+            .entityType(
+                pipelineName.getValue()) // this must match the configuration the method is set to
+            // launch with.  Do we want to modify the current method
+            // configuration each time so this matches?
+            .useCallCache(true)
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java:68]: do we want to populate these values with those in the imputation config?

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            .deleteIntermediateOutputFiles(false)
+            .useReferenceDisks(false)
+            .userComment("this is my flightid: %s".formatted(flightContext.getFlightId()))
+            .methodConfigurationNamespace(controlWorkspaceProject)
+            .methodConfigurationName(
+                "ImputationBeagleEmpty"); // need to not hardcode this eventually
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java:74]: are we able to import a WDL into a workspace with a custom name like we did in Azure? so like get the ImputationBeagleEmpty wdl but have it be called ImputationBeagle? I think this is possible (we'd need to create a separate workspace for it, like we did in Azure) and then we could have this come from the db like before

File: service/src/test/java/bio/terra/pipelines/dependencies/rawls/RawlsServiceTest.java [hunk 5]
```diff
@@ -24,13 +33,159 @@ class RawlsServiceTest extends BaseEmbeddedDbTest {
...
+  }
+
+  @Test
+  void getSubmissionStatus() throws Exception {
+    UUID expectedUUID = UUID.randomUUID();
+    Submission expectedResponse =
+        new Submission()
+            .status(SubmissionStatus.SUBMITTING)
+            .submissionId(expectedUUID.toString())
+            .cost(3.45F);
```
Comment (mmorgantaylor) [inline service/src/test/java/bio/terra/pipelines/dependencies/rawls/RawlsServiceTest.java:139]: oh cool that cost comes back. probably not in any timely way for us to do anything with it, right?

**Review by mmorgantaylor — COMMENTED** [review]
bunch of questions to talk about

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/AddDataTableRowStep.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    String controlWorkspaceProject =
+        inputParameters.get(RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_PROJECT, String.class);
+    PipelinesEnum pipelineName =
+        inputParameters.get(JobMapKeys.PIPELINE_NAME.getKeyName(), PipelinesEnum.class);
+
+    Entity entity =
+        new Entity()
+            .entityType(pipelineName.getValue())
+            .name(flightContext.getFlightId())
+            .putAttributesItem(
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/AddDataTableRowStep.java:56]: populate this with ALL_PIPELINE_INPUTS from prepareimputationinputs

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import bio.terra.pipelines.dependencies.rawls.RawlsServiceApiException;
+import bio.terra.pipelines.dependencies.sam.SamService;
+import bio.terra.pipelines.dependencies.stairway.JobMapKeys;
+import bio.terra.pipelines.stairway.imputation.RunImputationJobFlightMapKeys;
+import bio.terra.rawls.model.SubmissionReport;
+import bio.terra.rawls.model.SubmissionRequest;
+import bio.terra.stairway.*;
+
+/**
+ * This step submits a run set to cromwell using the rawls submission endpoint.
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java:15]: update to include stuff about imputationconfig being used

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    // validate and extract parameters from working map
+    FlightMap workingMap = flightContext.getWorkingMap();
+
+    // create submission request
+    SubmissionRequest submissionRequest =
+        new SubmissionRequest()
+            .entityName(flightContext.getFlightId())
+            .entityType(
+                pipelineName.getValue()) // this must match the configuration the method is set to
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java:58]: https://broadworkbench.atlassian.net/browse/TSPS-301 to address this

File: service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            // launch with.  Do we want to modify the current method
+            // configuration each time so this matches?
+            .useCallCache(true)
+            .deleteIntermediateOutputFiles(false)
+            .useReferenceDisks(false)
+            .userComment("this is my flightid: %s".formatted(flightContext.getFlightId()))
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/SubmitCromwellRunSetStep.java:64]: update comment to match azure or just make it good

**Review by mmorgantaylor — APPROVED** [review]
awesome! a couple super minor comment suggestions

**Pull request merged**